File: Homework/Hoang_Nguyen/lesson9/285_inorderSuccessorInBST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @return {TreeNode}
 */
var inorderSuccessor = function (root, p) {
  let ans = null;

  while (root) {
    if (root.val <= p.val) {
      root = root.right;
    } else {
      ans = root;
      root = root.left;
    }
  }

  return ans;
};

File: Homework/Ngan_Luu/285-inorder-successor-in-bst.py
# https://leetcode.com/problems/inorder-successor-in-bst/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def inorderSuccessor(self, root: TreeNode, p: TreeNode) -> Optional[TreeNode]:
        arr=[]
        self.res=None
        preNode=[None]
        def dfs(node):
            if not node:
                return
            dfs(node.left)
            if preNode[-1]==p.val:
                self.res=node
            preNode[-1]=node.val
            arr.append(node.val)
            dfs(node.right)
        dfs(root)
        return self.res